gnovm/pkg/gnolang/symbols_table.go

```diff
@@ -0,0 +1,36 @@
+package gnolang
+
+type SymbolTable struct {
+	scopes []*Scope
+}
+
+type Scope struct {
+	symbols map[string]struct{}
+}
+
+func NewSymbolTable() *SymbolTable {
+	return &SymbolTable{
+		scopes: []*Scope{newScope()},
+	}
+}
+
+func newScope() *Scope {
+	return &Scope{
+		symbols: make(map[string]struct{}),
+	}
+}
+
+func (st *SymbolTable) AddIdentifier(name string) {
+	if len(st.scopes) > 0 {
+		st.scopes[len(st.scopes)-1].symbols[name] = struct{}{}
+	}
+}
+
+func (st *SymbolTable) IdentifierExists(name string) bool {
+	for i := len(st.scopes) - 1; i >= 0; i-- {
```

Comment:
why this is needed to be backwards?

Reply:
Because thats how scopes work. You start with your current scope and go backwards until you find the identifier or don't find it at all.